### PR TITLE
🧭 fix: Handle Mistyped Handoff Tools Locally

### LINKED ISSUE #60: Handoff tool not found: wrong tool name (duplicated suffix) when transferring to agent
### Problem

When an agent that performs handoffs (in our case a router agent we created) tries to transfer to another agent (e.g. a “Code Refactorer”), it sometimes calls the handoff tool with a **wrong name**: the correct ID suffix is duplicated (e.g. `...7yu` becomes `...7yu7yu`). The tool is then not found on the host (handoff tools are graph-managed, not in the host’s `loadTools`), which leads to repeated failed handoffs and multiple “Transferred to Agent” messages without an actual transfer.

**Observed:** Several “Transferred to Agent” lines in a row, then the assistant reports that the transfer to the code refactorer failed.

**Provider/Model:** *Scaleway / devstral-2-123b-instruct-2512*

(The full ModelSpec configuration can be viewed [here](https://github.com/faktenforum/ai-chat-interface/blob/223fe85b21bc53a552479421f23e64c46d016bd1/packages/librechat-init/config/librechat.yaml#L482) and the full Agents configuration [here](https://github.com/faktenforum/ai-chat-interface/blob/223fe85b21bc53a552479421f23e64c46d016bd1/packages/librechat-init/config/agents.yaml#L868).)

### Relevant logs

```
2026-02-17 13:41:29 warn: [ON_TOOL_EXECUTE] Tool "lc_transfer_to_agent_V_29lG4o1TnQ3-C0wu7yu7yu" not found. Available: 

2026-02-17 13:41:31 warn: [ON_TOOL_EXECUTE] Tool "lc_transfer_to_agent_V_29lG4o1TnQ3-C0wu7yu7yu" not found. Available: 

2026-02-17 13:41:32 warn: [ON_TOOL_EXECUTE] Tool "lc_transfer_to_agent_V_29lG4o1TnQ3-C0wu7yu7yu" not found. Available: 

2026-02-17 13:41:33 warn: [ON_TOOL_EXECUTE] Tool "lc_transfer_to_agent_V_29lG4o1TnQ3-C0wu7yu7yu" not found. Available: 

2026-02-17 13:41:34 warn: [ON_TOOL_EXECUTE] Tool "lc_transfer_to_agent_V_29lG4o1TnQ3-C0wu7yu7yu" not found. Available: 

2026-02-17 13:41:36 error: [api/server/controllers/agents/client.js #sendCompletion] Operation aborted 400 Cannot set `add_generation_prompt` to True when the last message is from the assistant. Consider using `continue_final_message` instead.

2026-02-17 13:41:36 error: [api/server/controllers/agents/client.js #sendCompletion] Unhandled error type 400 Cannot set `add_generation_prompt` to True when the last message is from the assistant. Consider using `continue_final_message` instead.
```

The correct tool name would be `lc_transfer_to_agent_V_29lG4o1TnQ3-C0wu7yu` (without the duplicated `7yu`).

### Screenshot

Image

### Context

- Handoff tools are created with names like `lc_transfer_to_` + `edge.to` (agent ID). Edges/init do not duplicate the ID.
- Possible mitigation: when a handoff tool name is “wrong” but prefix-matches a known handoff tool, return a ToolMessage to the LLM with the correct name so it can call the right tool in the next turn (instead of executing silently or leaving the host to report “tool not found”).

### PR DESCRIPTION
**Pull request by danny-avila**
## Summary

I fixed mistyped graph-managed handoff tool calls so they fail closed inside `ToolNode` with exact-name guidance instead of being silently executed as a guessed handoff or dispatched to the host tool executor.

Fixes #60.

- Added exact handoff-name detection helpers that only suggest a registered handoff when the model emitted a strict longer prefix match.
- Routed unknown handoff-shaped calls through the local/direct `runTool` error path so the model receives a normal `ToolMessage` error tied to its original tool call id.
- Preserved host event dispatch for real event-driven tools, while preventing typo-shaped handoffs from reaching `ON_TOOL_EXECUTE`.
- Added regression coverage for plain graph handoffs, mixed event-driven host tools, local-execution direct-tool overlays, and event-driven ToolNodes without direct tool names.
- Added an opt-in live Anthropic handoff integration test using real `Run.processStream` execution and a reusable `test:live:handoffs` script.

This is empirically better than the contributor patch because the tests now measure the behavioral contract directly: a mistyped handoff is not executed, does not reach the host executor, and still returns actionable exact-name feedback. The live integration test also proves a real Anthropic handoff routes through the graph-managed transfer tool and preserves instructions into the receiving agent. The prior approach added synthetic correction output after resolving the typo, but it introduced several extra passes over the same call list and did not prove the host-dispatch leakage case. This patch keeps the error in the same existing unknown-tool path and verifies the real edge cases with executable assertions.

## Change Type

- [x] Bug fix (non-breaking change which fixes an issue)

## Testing

- Ran `NODE_OPTIONS='--experimental-vm-modules' npx jest src/specs/agent-handoffs.test.ts --runInBand`.
- Ran `NODE_OPTIONS='--experimental-vm-modules' npx jest src/specs/agent-handoffs.live.test.ts --runInBand` to verify the live test is skipped by default.
- Ran `NODE_OPTIONS='--experimental-vm-modules' npx jest src/specs/agent-handoffs.live.test.ts src/specs/agent-handoffs.test.ts --runInBand`.
- Ran `npm run test:live:handoffs` with `.env` credentials; the live Anthropic handoff passed with `claude-sonnet-4-6`.
- Ran `npx tsc --noEmit`.
- Ran `npx eslint src/tools/ToolNode.ts src/specs/agent-handoffs.test.ts`.
- Ran `npx eslint src/specs/agent-handoffs.live.test.ts`.
- Ran `git diff --check`.
- Ran live OpenAI handoff steering script successfully across three real handoff scenarios.
- Ran a compact live Anthropic handoff smoke with `claude-sonnet-4-6`; it returned `handoff: true`, `toolNames: ["lc_transfer_to_specialist"]`, and `specialistReplied: true`.

### **Test Configuration**:

- Base branch: `origin/dev` at `da06b06ada4b6fe3c12757fa2084b445e84b7197`
- Node: `v20.19.5`
- npm: `10.8.2`
- Anthropic model used for live smoke and live Jest handoff: `claude-sonnet-4-6`

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] My changes do not introduce new warnings
- [x] I have written tests demonstrating that my changes are effective or that my feature works
- [x] Local unit tests pass with my changes

### CODE REVIEW

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
@codex review

File: src/tools/ToolNode.ts [hunk 1]
```diff
@@ -581,6 +585,61 @@ export class ToolNode<T = any> extends RunnableCallable<T, T> {
     return this.fileCheckpointer;
   }
 
+  private getRegisteredHandoffNames(): Iterable<string> {
+    return this.directToolNames != null && this.directToolNames.size > 0
+      ? this.directToolNames
+      : this.toolMap.keys();
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/ToolNode.ts]: Addressed in `1c99eae`: `getRegisteredHandoffNames()` now treats `directToolNames` as an overlay and includes `toolMap` handoff names as well. I added coverage for a handoff in `toolMap` with an unrelated direct local tool name present, and verified with `NODE_OPTIONS='--experimental-vm-modules' npx jest src/specs/agent-handoffs.test.ts --runInBand`, `npx tsc --noEmit`, `npx eslint src/tools/ToolNode.ts src/specs/agent-handoffs.test.ts`, and `git diff --check`.

File: src/tools/ToolNode.ts [hunk 2]
```diff
@@ -2849,8 +2919,9 @@ export class ToolNode<T = any> extends RunnableCallable<T, T> {
 
       if (this.eventDrivenMode && filteredCalls.length > 0) {
         const filteredIndices = filteredCalls.map((_, idx) => idx);
+        const directToolNames = this.directToolNames;
 
-        if (!this.directToolNames || this.directToolNames.size === 0) {
+        if (directToolNames == null || directToolNames.size === 0) {
           return this.executeViaEvent(filteredCalls, config, input, {
             batchIndices: filteredIndices,
             turn,
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/ToolNode.ts]: Addressed in `7c074e9`: the event-driven batch path now classifies calls before using the all-host fast path, so unknown handoff-shaped calls stay local even when `directToolNames` is empty. I added a regression for an event-driven `ToolNode` with no direct tool names, a real host tool, and a mistyped handoff that exists only in `toolMap`. Verified with `NODE_OPTIONS='--experimental-vm-modules' npx jest src/specs/agent-handoffs.test.ts --runInBand`, `npx tsc --noEmit`, `npx eslint src/tools/ToolNode.ts src/specs/agent-handoffs.test.ts`, and `git diff --check`.

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
@codex review
